riscv-executor/src/lib.rs

```diff
@@ -30,62 +34,105 @@ pub mod poseidon_gl;
 /// TODO: get this value from some authoritative place
 const PC_INITIAL_VAL: usize = 2;
 
-#[derive(Clone, Copy, PartialEq, Eq)]
-pub struct Elem(pub i64);
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Elem<F: FieldElement> {
+    /// Only the ranges of i32 and u32 are actually valid for a Binary value.
+    /// I.e., [-2**31, 2**32).
+    Binary(i64),
+    Field(F),
+}
 
-impl Elem {
-    fn u(&self) -> u32 {
-        self.0.try_into().unwrap()
+impl<F: FieldElement> Elem<F> {
+    /// Interprets the value of a field as a binary, if it can be represented either as a
+    /// u32 or a i32.
+    ///
+    /// Panics otherwise.
+    pub fn new_from_fe_as_bin(value: &F) -> Self {
+        if let Some(v) = value.to_integer().try_into_u32() {
+            Self::Binary(v as i64)
+        } else if let Some(v) = value.try_into_i32() {
+            Self::Binary(v as i64)
```

Comment:
if the input is F is fits into i32, why wouldn't it have fit u32 above?

Reply:
If it is negative, it doesn't match `u32`, and if it is >= 2**31, it doesn't match `i32`.

Reply:
but if the input is F, how is it negative?